### PR TITLE
Improve performance of index_select by avoiding item

### LINKED ISSUE #61788: [feature request] index_select is very slow on sparse tensors (and my proposed algorithm to fix it)
torch.index_select is supposed to work on both dense, and sparse tensors. For dense tensors it's pretty amazing, but for sparse tensors it's painfully slow. Here's an example I ran in a jupyter notebook that shows this:
```
import torch
from earth_mesh.entity.sparse_tensor import SparseTensor 
# SparseTensor is a class that has some custom functions I wrote that operate on sparse tensors

dense_range = torch.arange(100 * 100 * 100).view(100, 100, 100)
sparse_range = dense_range.to_sparse()

indices = torch.randint(0, 100, (100,))

%time torch.index_select(dense_range, 0, indices)
%time torch.index_select(sparse_range, 0, indices)
%time SparseTensor._index_select(SparseTensor(sparse_range), 0, indices)
```
This had the result:

> CPU times: user 12.5 ms, sys: 0 ns, total: 12.5 ms
Wall time: 513 µs
CPU times: user 26.9 s, sys: 5.26 ms, total: 26.9 s
Wall time: 26.8 s
CPU times: user 254 ms, sys: 47.8 ms, total: 301 ms
Wall time: 83.8 ms

As you can see index_select for a dense tensor runs incredibly quickly, but the pytorch index_select is atrociously slow for sparse tensors. However, the algorithm I wrote was several hundred times faster. For the use case I originally built this for it was several thousand times faster! I was wondering if my algorithm could replace the current one that computes index_select for sparse tensors.

The only catch is that the space complexity isn't amazing. If n=len(indices) and m=number of occurrences of most common index in tensor._indices()[dim] then the space complexity is O(nm). This might seem bad, but this is also the space complexity of the size of the output! (This is true, but in my algorithm the best case is very close to the worst case, which I'm sure isn't true of the current index_select)

My function is as follows

```
@classmethod
def _index_select(cls, tensor, dim, indices):
    is_sorted = torch.all(torch.diff(tensor._indices()[dim]) >= 0)
    if not is_sorted:
        sort_indices = torch.argsort(tensor._indices()[dim])
        tensor = torch.sparse_coo_tensor(tensor._indices()[:, sort_indices], tensor._values()[sort_indices], tensor.shape)

    search_result_left = torch.searchsorted(tensor._indices()[dim], indices, right=False)
    search_result_right = torch.searchsorted(tensor._indices()[dim], indices, right=True)
    num_results = search_result_right - search_result_left

    index_mesh_grid = torch.stack(torch.meshgrid(
        torch.arange(torch.max(num_results)),
        torch.arange(indices.shape[0])
    ))

    desired_index_info = torch.where(
        index_mesh_grid[0] < num_results,
        index_mesh_grid,
        torch.full_like(index_mesh_grid, -1)
    ).transpose(1, 2)

    filtered_index_info = desired_index_info[desired_index_info != -1].view(2, -1)

    tensor_indices = search_result_left[filtered_index_info[1]] + filtered_index_info[0]
    return_index = tensor._indices()[:, tensor_indices]
    return_index[dim] = filtered_index_info[1]
    new_shape = *tensor.shape[:dim], indices.shape[0], *tensor.shape[dim + 1:]
    return torch.sparse_coo_tensor(return_index, tensor._values()[tensor_indices], new_shape)

```
If you're interested in the algorithm but it's a bit too dense to understand then I'm more than happy to explain it!

cc @VitalyFedyunin @ngimel @heitorschueroff @nikitaved @pearu @cpuhrsch @IvanYashchuk

### PR DESCRIPTION
**Pull request by cpuhrsch**
Partially fixes #61788

From a CUDA perspective: item already pulls all Tensor content onto the host (albeit one-by-one), which incurs very expensive memory transfers. This way we'll do it all at once.
From a CPU perspective: item has a lot of overhead as a native function in comparison to simply using a pointer.

Overall there's still lots of performance gains to be had, but this is a small change that should take us into a more usable landscape. This doesn't land a separate benchmark, but I postulate that's not necessary to decide on the benefit of this (we'll also see if it shows up indirectly), however is still a good follow-up item. 

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Benchmark script
```
import torch
import time

def benchmark_torch_function(iters, f, *args, **kwargs):
    f(*args, **kwargs)
    if torch.cuda.is_available():
        torch.cuda.synchronize()
        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)
        start_event.record()
    else:
        t0 = time.time()
    for i in range(iters):
        f(*args, **kwargs)
    if torch.cuda.is_available():
        end_event.record()
        torch.cuda.synchronize()
        return start_event.elapsed_time(end_event)
    else:
        return (time.time() - t0)

def timing(device):
    dense_range = torch.arange(100 * 100 * 100).view(100, 100, 100).to(device)
    sparse_range = dense_range.to_sparse().to(device)
    indices = torch.randint(0, 100, (100,)).to(device)
    iters = 10
    dense = benchmark_torch_function(iters, torch.index_select, dense_range, 0, indices)
    sparse = benchmark_torch_function(iters, torch.index_select, sparse_range, 0, indices)
    print(f"device: {str(device).rjust(4)}, dense: {dense/iters:4.2f}ms, sparse: {sparse/iters:4.2f}ms")

timing(torch.device('cpu'))
timing(torch.device('cuda'))
```

Baseline

Branch
```
device:  cpu, dense: 0.73ms, sparse: 245.07ms
device: cuda, dense: 0.03ms, sparse: 230.86ms
```

**Comment by cpuhrsch** [comment]
@IAmKohlton could I kindly ask you to patch and verify the results of this PR for your usecases?

**Comment by IAmKohlton** [comment]
@cpuhrsch yeah, I can try do that in the next few days, and get back to you about it

**Review by pearu — APPROVED** [review]
LGTM!

**Comment by cpuhrsch** [comment]
Note that this requires ```indices[dim]``` and ```index``` to be contiguous, so this change is introducing a bug that our tests don't catch. I'll update the tests to detect the bug and then fix this issue.
